aws/service.ts

```diff
@@ -247,74 +247,109 @@ function computeImage(
     }
     else if (container.image) {
-        return pulumi.all(preEnv).apply(e => ({ image: imageName, environment: e }));
+        return computeImageFromImage(preEnv, imageName);
     }
     else if (container.function) {
-        const func = container.function;
-        preEnv.PULUMI_SRC = pulumi.runtime.serializeFunctionAsync(func);
-
-        // TODO[pulumi/pulumi-cloud#85]: Put this in a real Pulumi-owned Docker image.
-        // TODO[pulumi/pulumi-cloud#86]: Pass the full local zipped folder through to the container (via S3?)
-        return pulumi.all(preEnv).apply(e => ({ image: imageName, environment: e }));
+        return computeImageFromFunction(container.function, preEnv, imageName);
     }
     else {
         throw new RunError("Invalid container definition: `image`, `build`, or `function` must be provided");
     }
 }
 
+function computeImageFromBuild(
+        preEnv: Record<string, pulumi.Input<string>>,
+        build: string | cloud.ContainerBuild,
+        imageName: string,
+        repository: aws.ecr.Repository): pulumi.Output<ImageOptions> {
+
+    // This is a container to build; produce a name, either user-specified or auto-computed.
+    pulumi.log.debug(`Building container image at '${build}'`, repository);
+    const { repositoryUrl, registryId } = repository;
+
+    return pulumi.all([repositoryUrl, registryId]).apply(([repoUrl, regId]) =>
+        computeImageFromBuildWorker(preEnv, build, imageName, repoUrl, regId, repository));
+}
+
+function computeImageFromBuildWorker(
+        preEnv: Record<string, pulumi.Input<string>>,
+        build: string | cloud.ContainerBuild,
+        imageName: string,
+        repositoryUrl: string,
+        registryId: string,
+        logResource: pulumi.Resource): pulumi.Output<ImageOptions> {
+
+    let imageDigest: pulumi.Output<string>;
+    // See if we've already built this.
+    if (imageName && buildImageCache.has(imageName)) {
+        // We got a cache hit, simply reuse the existing digest.
+        // Safe to ! the result since we checked buildImageCache.has above.
+        imageDigest = buildImageCache.get(imageName)!;
+        imageDigest.apply(d =>
+            pulumi.log.debug(`    already built: ${imageName} (${d})`, logResource));
+    } else {
+        // If we haven't, build and push the local build context to the ECR repository, wait for
+        // that to complete, then return the image name pointing to the ECT repository along
+        // with an environment variable for the image digest to ensure the TaskDefinition get's
+        // replaced IFF the built image changes.
+        imageDigest = docker.buildAndPushImage(imageName, build, repositoryUrl, logResource, async () => {
+            // Construct Docker registry auth data by getting the short-lived authorizationToken from ECR, and
+            // extracting the username/password pair after base64-decoding the token.
+            //
```

Comment:
the important line is 5 lines below (in teh collapsed section).  We call `await aws.ecr.getCredentials({ registryId: registryId });`.  Now that registryId is passed in as string, there's no problem with passing Inputs to that function.